File: sub_pub/adapters/kafka.py
"""Kafka adapter"""
import logging
from typing import Iterator, List, Optional

from sub_pub.core.interfaces import MessageSource, MessagePublisher
from sub_pub.core.message import Message

logger = logging.getLogger(__name__)


class KafkaSource(MessageSource):
    """Kafka message source"""
    
    def __init__(self, config: dict):
        self.config = config
        self.consumer = None
        
    def connect(self) -> None:
        """Establish connection to Kafka"""
        try:
            from kafka import KafkaConsumer
            
            self.consumer = KafkaConsumer(
                **self.config,
                enable_auto_commit=False
            )
            logger.info("Connected to Kafka source")
        except ImportError:
            logger.error("kafka-python package not installed. Install with: pip install kafka-python")
            raise
        
    def subscribe(self, topics: List[str]) -> None:
        """Subscribe to topics"""
        if self.consumer:
            self.consumer.subscribe(topics)
            logger.info(f"Subscribed to Kafka topics: {topics}")
        
    def consume(self) -> Iterator[Message]:
        """Consume messages from Kafka"""
        if not self.consumer:
            raise RuntimeError("Consumer not connected")
        
        for record in self.consumer:
            headers = {k: v.decode('utf-8') if isinstance(v, bytes) else str(v) 
                      for k, v in (record.headers or [])}
            
            yield Message(
                payload=record.value,
                headers=headers,
                topic=record.topic,
                key=record.key.decode('utf-8') if record.key else None,
                partition=record.partition,
                offset=record.offset,
                timestamp=record.timestamp
            )
        
    def close(self) -> None:
        """Close Kafka connection"""
        if self.consumer:
            self.consumer.close()
            logger.info("Kafka source closed")
        
    def commit(self, message: Optional[Message] = None) -> None:
        """Commit message offset"""
        if self.consumer:
            self.consumer.commit()


class KafkaPublisher(MessagePublisher):
    """Kafka message publisher"""
    
    def __init__(self, config: dict):
        self.config = config
        self.producer = None
        
    def connect(self) -> None:
        """Establish connection to Kafka"""
        try:
            from kafka import KafkaProducer
            
            self.producer = KafkaProducer(**self.config)
            logger.info("Connected to Kafka publisher")
        except ImportError:
            logger.error("kafka-python package not installed. Install with: pip install kafka-python")
            raise
        
    def publish(self, message: Message, topic: str) -> None:
        """Publish a message to Kafka"""
        if not self.producer:
            raise RuntimeError("Producer not connected")
        
        headers = [(k, v.encode('utf-8')) for k, v in message.headers.items()]
        key = message.key.encode('utf-8') if message.key else None
        
        self.producer.send(
            topic,
            value=message.payload,
            key=key,
            headers=headers
        )
        
    def flush(self) -> None:
        """Flush buffered messages"""
        if self.producer:
            self.producer.flush()
        
    def close(self) -> None:
        """Close Kafka connection"""
        if self.producer:
            self.producer.close()
            logger.info("Kafka publisher closed")

File: rust/src/adapters/kafka.rs
use async_trait::async_trait;

use crate::core::{Message, MessagePublisher, MessageSource};
use crate::error::SubPubError;

/// Stub Kafka source.  Requires the `kafka` feature to be fully implemented.
pub struct KafkaSource {
    pub bootstrap_servers: String,
    pub group_id: String,
    pub topics: Vec<String>,
}

impl KafkaSource {
    pub fn new(bootstrap_servers: impl Into<String>, group_id: impl Into<String>) -> Self {
        Self {
            bootstrap_servers: bootstrap_servers.into(),
            group_id: group_id.into(),
            topics: Vec::new(),
        }
    }
}

#[async_trait]
impl MessageSource for KafkaSource {
    async fn connect(&mut self) -> Result<(), SubPubError> {
        tracing::warn!(
            "KafkaSource: full implementation pending (enable the `kafka` feature flag)"
        );
        Ok(())
    }

    async fn subscribe(&mut self, topics: &[&str]) -> Result<(), SubPubError> {
        self.topics = topics.iter().map(|t| t.to_string()).collect();
        Ok(())
    }

    async fn consume(&mut self) -> Result<Option<Message>, SubPubError> {
        Err(SubPubError::Adapter(
            "KafkaSource: full implementation pending".to_string(),
        ))
    }

    async fn commit(&mut self, _message: &Message) -> Result<(), SubPubError> {
        Ok(())
    }

    async fn close(&mut self) -> Result<(), SubPubError> {
        Ok(())
    }
}

/// Stub Kafka publisher.
pub struct KafkaPublisher {
    pub bootstrap_servers: String,
}

impl KafkaPublisher {
    pub fn new(bootstrap_servers: impl Into<String>) -> Self {
        Self {
            bootstrap_servers: bootstrap_servers.into(),
        }
    }
}

#[async_trait]
impl MessagePublisher for KafkaPublisher {
    async fn connect(&mut self) -> Result<(), SubPubError> {
        tracing::warn!(
            "KafkaPublisher: full implementation pending (enable the `kafka` feature flag)"
        );
        Ok(())
    }

    async fn publish(&mut self, _message: &Message, _topic: &str) -> Result<(), SubPubError> {
        Err(SubPubError::Adapter(
            "KafkaPublisher: full implementation pending".to_string(),
        ))
    }

    async fn flush(&mut self) -> Result<(), SubPubError> {
        Ok(())
    }

    async fn close(&mut self) -> Result<(), SubPubError> {
        Ok(())
    }
}
